README.txt
SlideNote: A jQuery plugin for flexible, customizable sliding notifications.

HOW TO USE

1.	Include SlideNote in the header of your page. Make sure it is included after
	jQuery
	
	<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.4.2/jquery.min.js" type="text/javascript"></script>
	<script src="jquery.slidenote.js" type="text/javascript"></script>
	
2.	Apply the plugin to a single or set of elements.

	$('#myNote').slideNote();	// single element
	$('.notes').slideNote();	// set of elements

3. 	Pass options to the slideNote() function to modify the behavior:

	$('#myNote').slideNote({	
		where: 640,
		corner: 'left',
		url: 'ajax.html',
		container: 'note',
		closeImage: '/images/close.png'
	});
	
OPTIONS

where 			specified how far down the user must scroll before the notification
						slides into view
		
corner			controls from which side of the screen the notification will slide in

URL					tells SlideNote from where to pull the data to use as the content of
						the notification. if no container is specified, the entire page is used
		
container		to be used in conjunction with 'URL.' when specified, only pulls
						the element having the specified ID from the specified URL.
					
closeImage	displays the image at the specified URL. When clicked, it will slide the 
						note out of view. The note will not display until the next time the user 
						scrolls passed the <strong>where</strong> value.

CONTACT ME

Website: http://tommcfarlin.com
Email:	 tom@slidenote.info

changelog.txt
1.0		9/29/2010		initial release


tests/test.js
/*
 * SlideNote Test Suite
 * Unit Tests for SlideNote
 * A JQuery Plugin for flexible, customizable sliding notifications.
 *
 * Copyright 2010 Tom McFarlin, http://tommcfarlin.com
 * Released under the MIT License
 * More information: http://slidenote.info
*/

/*-------------------------------------------------------*
 * Initialization & Setup
 *-------------------------------------------------------*/
 
module("Initialization &amp; Setup");
test("Setup", function() {
	function defined() {
		return $("#note").length;
	}
	equal(defined(), 1, 'A single instance of the note container is found.');
});

test('Initialization', function() {
	function init() {
		$('#note').slideNote();
		return $('#note').css('position') === 'fixed' && $('#note').css('bottom') === '0px';
	}
	equal(init(), true, 'Verifies that the initial CSS position and bottom values are properly set.');
});

/*-------------------------------------------------------*
 * Default Option Values
 *-------------------------------------------------------*/

module("Default Option Values");
test("where", function() {
	function defaultWhereValue() {
		return $.fn.slideNote.defaults.where
	}
	equal(defaultWhereValue(), 640, 'At what scroll position the note will be displayed if no value has been provided.');
});

test("corner", function() {
	function defaultCornerValue() {
		return $.fn.slideNote.defaults.corner
	}
	equal(defaultCornerValue(), 'right', 'The corner in which the note will appear if no value has been specified.');
});

test("URL", function() {
	function defaultUrlValue() {
		return $.fn.slideNote.defaults.url
	}
	equal(defaultUrlValue(), null, 'Verifies the URL field is null when not specified.');
});

test("container", function() {
	function defaultUrlValue() {
		return $.fn.slideNote.defaults.container
	}
	equal(defaultUrlValue(), '', 'The default container value if no asynchronous element ID is provided.');
});

/*-------------------------------------------------------*
 * Event Tests
 *-------------------------------------------------------*/

module("Event Tests");
asyncTest("slideIn", function() {
	setTimeout(function() {
		$('#note').slideNote().trigger('slideIn');
		start();
		equal($('#note').is(':visible'), true, 'Verifies that the note is visible after the slideIn animation has fired.');
	}, 1000);
});

asyncTest("slideOut", function() {
	setTimeout(function() {
		// This test needs some work as its based on CSS and not the true events.
		$('#note').slideNote()
			.css('display', 'block')
			.trigger('slideOut');
		start();
		equal($('#note').css('display') === 'none', false, 'Verifies that the note is invisible after the slideOut animation has fired.');
	}, 3000);
});

/*-------------------------------------------------------*
 * Ajax Tests
 *-------------------------------------------------------*/
module("Ajax Tests");
asyncTest("retrieveData", function() {
	$('#ajaxNote').slideNote({
		url: 'ajax.html'
	});
	setTimeout(function() {
		equal($('#ajaxNote').children(':first').attr('id') === 'container', true, 'Verifies that the entire page is properly appended to the DOM.');
		start();
	}, 1000);
});

asyncTest("retrieveSpecificData", function() {
	$('#ajaxNote').slideNote({
		url: 'ajax2.html',
		container: 'container2'
	});
	setTimeout(function() {
		equal($('#ajaxNote').children(':first').attr('id') === 'container2', true, 'Verifies that the specified element is properly appended to the DOM.');
		start();
	}, 1000);
});

js/jquery.slidenote.js
/*
 * SlideNote
 * A jQuery Plugin for flexible, customizable sliding notifications.
 *
 * Copyright 2010 Tom McFarlin, http://tommcfarlin.com
 * Released under the MIT License
 * More information: http://slidenote.info
*/

(function($) {

	$.slideNoteCount = 0;
	
	$.fn.slideNote = function(options) {
		
		$.slideNoteCount += this.length;
		
		var opts = $.extend({}, $.fn.slideNote.defaults, $.fn.slideNote.private, options);
		return this.each(function() {
			
			var $note = _init(this, opts);
		
			var oDoc = $.browser.msie === true ? window : document;
			$(oDoc).scroll(function() {
				if($(this).scrollTop() === 0) {
					opts.bIsClosed = false;
				}
				if($(this).scrollTop() > opts.where) {
					if(!$note.is(':visible') && !opts.bIsClosed) {
						$note.trigger('slideIn');
					}
				} else if ($(this).scrollTop() < opts.where && $note.queue('fx')[0] !== 'inprogress') {
					if($note.is(':visible')) {
						$note.trigger('slideOut');
					}
				}
			});
			
		});
		
	};
	
	function _init(obj, opts) {
	
		$(obj).toggle()
			.css(opts.corner, -1 * $(obj).outerWidth())
			.css({
				'position': 'fixed',
				'bottom': 0
			})
			.bind('slideIn', function(evt) {
				_slideIn(evt, obj, opts);
			})	
			.bind('slideOut', function(evt) {
				_slideOut(evt, obj, opts);
			});
		
		_retrieveData(obj, opts);
		_addCloseImage(obj, opts);
		
		return $(obj);
		
	}
	
	function _slideIn(evt, obj, opts) {	
		var direction = opts.corner === 'right' ? { 'right' : 0 } : { 'left' : 0 } ;
		$(obj).show().animate(direction, 1000, 'swing');
	}
	
	function _slideOut(evt, obj, opts) {
		
		var direction = opts.corner === 'right' ? { 'right' : -1 * $(obj).outerWidth() } : { 'left' : -1 * $(obj).outerWidth() };
		$(obj).animate(direction, 1000, 'swing', function() {
			if($.slideNoteCount === 1) {
				$(obj).stop(true).hide();
			} else {
				$(obj).hide();
			}
		});
	
		if(opts.closeImage !== null && evt.target.id === $(obj).attr('id') + '_close') {
			opts.bIsClosed = true;
		}
		
	}
	
	function _retrieveData(obj, opts) {
		if(opts.url !== null) {
			if(opts.container.length !== 0 && opts.container.indexOf('#') === -1) {
				opts.container = '#' + opts.container;
			}
			
			var sUrl = opts.container.length === 0 ? opts.url : opts.url + ' ' + opts.container;
			$(obj).load(sUrl, function() {
				if(opts.closeImage !== null) {
					_addCloseImage(obj, opts);
				}
			});
		}
	}
	
	function _addCloseImage(obj, opts) {
		if(opts.closeImage !== null) {
			var oImg = document.createElement('img');
			$(oImg).attr('src', opts.closeImage)
						 .attr('alt', 'close')
						 .attr('id', $(obj).attr('id') + '_close')
						 .hover(function() {
								$(this).css('cursor', 'pointer');
						 }).click(function(evt) {
								evt.stopPropagation();
								$(this).trigger('slideOut');
						 });
			$(obj).prepend(oImg);
		}
	}

	$.fn.slideNote.defaults = {
		where: 640,
		corner: 'right',
		url: null,
		container: '',
		closeImage: null
	};
	
	$.fn.slideNote.private = {
		_bIsClosed: false
	}
	
})(jQuery);